src/threads_mngr.c

```diff
@@ -129,55 +123,45 @@ static int test_and_start(void) {
 __attribute__ ((noinline))
 static void invoke_callback(int sig) {
     UNUSED(sig);
-
-    /* If the lock is already locked for write, we are running cleanups, no reason to proceed. */
-    if(0 != pthread_rwlock_tryrdlock(&globals_rw_lock)) {
-        serverLog(LL_WARNING, "threads_mngr: ThreadsManager_cleanups() is in progress, can't invoke signal handler.");
-        return;
-    }
-
-    if (g_callback) {
-        g_callback();
-        size_t curr_done_count;
-        atomicIncrGet(g_num_threads_done, curr_done_count, 1);
-
-        /* last thread shuts down the light */
-        if (curr_done_count == g_tids_len) {
-            sem_post(&wait_for_threads_sem);
-        }
+    run_on_thread_cb callback = g_callback;
+    if (callback) {
+        callback();
+        atomicIncr(g_num_threads_done, 1);
+    } else {
+        serverLogFromHandler(LL_WARNING, "tid %ld: ThreadsManager g_callback is NULL", syscall(SYS_gettid));
     }
-
-    pthread_rwlock_unlock(&globals_rw_lock);
 }
 
 static void wait_threads(void) {
-    struct timespec ts;
-    clock_gettime(CLOCK_REALTIME, &ts);
+    struct timespec timeout_time;
+    clock_gettime(CLOCK_REALTIME, &timeout_time);
 
     /* calculate relative time until timeout */
-    ts.tv_sec += RUN_ON_THREADS_TIMEOUT;
+    timeout_time.tv_sec += RUN_ON_THREADS_TIMEOUT;
+
+    /* Wait until all threads are done to invoke the callback or until we reached the timeout */
+    size_t curr_done_count;
+    struct timespec curr_time;
 
-    int status = 0;
+    do {
+        atomicGet(g_num_threads_done, curr_done_count);
+        clock_gettime(CLOCK_REALTIME, &curr_time);
```

Comment:
maybe we should add a short sleep / yield here? just so that it's not an intense busy loop?

Reply:
i think that in most cases it will just increase this loop duration. I experimented with different timeouts and noticed that I need a super low timeout (in ns) to get even one of the threads to be late.

Reply:
still, on an overloaded system, not sure if the false system call (VDSO) will induce a context switch, rather than just eat CPU. i'd rather add a `usleep(10)`, or even 1, or alike.

Reply:
I hate to say it but `usleep()` is not listed among signal-safe functions. 

Maybe we can use `select()`:

```c

struct timeval tv = {
    .tv_sec = 0,
    .tv_usec = 10
};

select(0, NULL, NULL, NULL, &tv);
```

Reply:
ok